File: app/src/main/java/net/nashihara/naroureader/fragments/RankingRecyclerViewFragment.java
package net.nashihara.naroureader.fragments;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.nashihara.naroureader.R;
import net.nashihara.naroureader.adapters.NovelDetailRecyclerViewAdapter;
import net.nashihara.naroureader.presenter.RankingRecyclerPresenter;
import net.nashihara.naroureader.databinding.FragmentRankingRecyclerBinding;
import net.nashihara.naroureader.databinding.ItemRankingRecyclerBinding;
import net.nashihara.naroureader.entities.NovelItem;
import net.nashihara.naroureader.listeners.FragmentTransactionListener;
import net.nashihara.naroureader.utils.DownloadUtils;
import net.nashihara.naroureader.views.RankingRecyclerView;
import net.nashihara.naroureader.widgets.FilterDialogFragment;
import net.nashihara.naroureader.widgets.ListDailogFragment;
import net.nashihara.naroureader.widgets.NovelDownloadDialogFragment;
import net.nashihara.naroureader.widgets.OkCancelDialogFragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import narou4j.entities.Novel;
import narou4j.enums.NovelGenre;

public class RankingRecyclerViewFragment extends Fragment implements RankingRecyclerView {

    private static final String TAG = RankingRecyclerViewFragment.class.getSimpleName();

    private FragmentRankingRecyclerBinding binding;

    private Context context;

    private FragmentTransactionListener replaceListener;

    private ArrayList<NovelItem> allItems = new ArrayList<>();

    private static final String PARAM_TYPE = "rankingType";

    private RankingRecyclerPresenter controller;

    public static RankingRecyclerViewFragment newInstance(String type) {
        RankingRecyclerViewFragment fragment = new RankingRecyclerViewFragment();
        Bundle args = new Bundle();
        args.putString(PARAM_TYPE, type);
        fragment.setArguments(args);
        return fragment;
    }

    public RankingRecyclerViewFragment() {}

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
        replaceListener = (FragmentTransactionListener) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_ranking_recycler, container, false);

        controller = new RankingRecyclerPresenter(this);

        binding.fab.setOnClickListener(v -> setupFab());

        binding.recycler.setLayoutManager(new LinearLayoutManager(context));

        NovelDetailRecyclerViewAdapter adapter = setupRecyclerView();
        binding.recycler.setAdapter(adapter);

        Bundle args = getArguments();
        if (args != null) {
            String typeStr = args.getString(PARAM_TYPE, "");
            controller.fetchRanking(typeStr);
        }

        return binding.getRoot();
    }

    private void setupFab() {
        ArrayList<String> filters =
            new ArrayList<>(Arrays.asList(getResources().getStringArray(R.array.genres)));
        filters.add(0, "完結済み");

        final NovelGenre[] filterIds =NovelGenre.values();


        boolean[] checked = new boolean[filters.size()];
        checked[0] = false;
        for (int i = 1; i < checked.length; i++) {
            checked[i] = true;
        }
        FilterDialogFragment checkBoxDialog = FilterDialogFragment
            .newInstance("小説絞込み", filters.toArray(new String[0]), checked, true,
                new FilterDialogFragment.OnDialogButtonClickListener() {
                    @Override
                    public void onPositiveButton(int which, boolean[] itemChecked, String min, String max) {
                        controller.filterNovelRanking(allItems, filterIds, itemChecked, min, max);
                    }

                    @Override
                    public void onNeutralButton(int which) {

                        NovelDetailRecyclerViewAdapter adapter = (NovelDetailRecyclerViewAdapter) binding.recycler.getAdapter();
                        adapter.getList().clear();
                        adapter.getList().addAll(allItems);
                    }
                });
        checkBoxDialog.show(getFragmentManager(), "multiple");
    }

    private NovelDetailRecyclerViewAdapter setupRecyclerView() {
        NovelDetailRecyclerViewAdapter adapter = new NovelDetailRecyclerViewAdapter(context, false);
        adapter.setOnItemClickListener(new NovelDetailRecyclerViewAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position, ItemRankingRecyclerBinding itemBinding) {
                if (view.getId() != R.id.btn_expand) {
                    NovelItem item = ((NovelDetailRecyclerViewAdapter) binding.recycler.getAdapter()).getList().get(position);
                    replaceListener.replaceFragment(NovelTableRecyclerViewFragment.newInstance(item.getNovelDetail().getNcode()), item.getNovelDetail().getTitle(), item);
                    return;
                }

                if (itemBinding.allStory.getVisibility() == View.GONE) {
                    itemBinding.allStory.setVisibility(View.VISIBLE);
                    itemBinding.keyword.setVisibility(View.VISIBLE);
                    itemBinding.btnExpand.setImageResource(R.drawable.ic_expand_less_black_24dp);
                } else {
                    itemBinding.allStory.setVisibility(View.GONE);
                    itemBinding.keyword.setVisibility(View.GONE);
                    itemBinding.btnExpand.setImageResource(R.drawable.ic_expand_more_black_24dp);
                }
            }

            @Override
            public void onItemLongClick(View view, final int position, ItemRankingRecyclerBinding itemBinding) {

                NovelDetailRecyclerViewAdapter adapter = (NovelDetailRecyclerViewAdapter) binding.recycler.getAdapter();

                final NovelItem item = adapter.getList().get(position);
                String[] strings = new String[]
                    {"小説を読む", "ダウンロード", "ブラウザで小説ページを開く", "ブラウザで作者ページを開く"};
                ListDailogFragment listDialog =
                    ListDailogFragment.newInstance(item.getNovelDetail().getTitle(), strings,
                        (dialog, which) -> longClickListDialogListener(which, position, item));
                listDialog.show(getFragmentManager(), "list_dialog");
            }
        });

        return adapter;
    }

    private void longClickListDialogListener(int which, int position, NovelItem item) {
        switch (which) {
            case 0: {
                NovelItem item1 = ((NovelDetailRecyclerViewAdapter) binding.recycler.getAdapter()).getList().get(position);
                replaceListener.replaceFragment(NovelTableRecyclerViewFragment.newInstance(
                    item1.getNovelDetail().getNcode()), item1.getNovelDetail().getTitle(), item1);
                break;
            }
            case 1: {
                DownloadUtils downloadUtils = new DownloadUtils() {
                    @Override
                    public void onDownloadSuccess(NovelDownloadDialogFragment dialog, final Novel novel) {
                        dialog.dismiss();

                        OkCancelDialogFragment okCancelDialog =
                            OkCancelDialogFragment.newInstance("ダウンロード完了", "ダウンロードした小説を開きますか？",
                                (dialog1, which1) -> {
                                    dialog1.dismiss();
                                    if (OkCancelDialogFragment.OK == which1) {
                                        NovelItem novelItem = new NovelItem();
                                        novelItem.setNovelDetail(novel);
                                        replaceListener.replaceFragment(
                                            NovelTableRecyclerViewFragment.newInstance(novel.getNcode()),
                                            novelItem.getNovelDetail().getTitle(), novelItem);
                                    }
                                });
                        okCancelDialog.show(getFragmentManager(), "okcansel");
                    }

                    @Override
                    public void onDownloadError(NovelDownloadDialogFragment dialog) {
                        Log.d(TAG, "onDownloadError: ");
                        dialog.dismiss();
                    }
                };
                downloadUtils.novelDownlaod(item.getNovelDetail(), getFragmentManager(), context);
                break;
            }
            case 2: {
                String url = "http://ncode.syosetu.com/" + item.getNovelDetail().getNcode() + "/";
                Uri uri = Uri.parse(url);
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
                break;
            }
            case 3: {
                String url = "http://mypage.syosetu.com/" + item.getNovelDetail().getUserId() + "/";
                Uri uri = Uri.parse(url);
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
                break;
            }
        }
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    public Context getContext() {
        return context;
    }

    private void reload() {
        getFragmentManager().beginTransaction().detach(this).attach(this).commit();
    }

    private void onLoadError() {
        binding.progressBar.setVisibility(View.GONE);
        binding.btnReload.setVisibility(View.VISIBLE);
        binding.btnReload.setOnClickListener(v -> {
            binding.progressBar.setVisibility(View.VISIBLE);
            binding.btnReload.setVisibility(View.GONE);
            reload();
        });
    }

    @Override
    public void showRanking(List<NovelItem> novelItems) {
        if (allItems.isEmpty()) {
            allItems = new ArrayList<>(novelItems);
        }

        NovelDetailRecyclerViewAdapter adapter = (NovelDetailRecyclerViewAdapter) binding.recycler.getAdapter();
        adapter.clearData();
        adapter.addDataOf(novelItems);
        binding.progressBar.setVisibility(View.GONE);
        binding.recycler.setVisibility(View.VISIBLE);
    }

    @Override
    public void showError() {
        onLoadError();
    }
}

File: app/src/main/java/net/nashihara/naroureader/fragments/RankingRecyclerViewFragment.kt
package net.nashihara.naroureader.fragments

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.databinding.DataBindingUtil
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import net.nashihara.naroureader.R
import net.nashihara.naroureader.adapters.NovelDetailRecyclerViewAdapter
import net.nashihara.naroureader.presenter.RankingRecyclerPresenter
import net.nashihara.naroureader.databinding.FragmentRankingRecyclerBinding
import net.nashihara.naroureader.databinding.ItemRankingRecyclerBinding
import net.nashihara.naroureader.entities.NovelItem
import net.nashihara.naroureader.listeners.FragmentTransactionListener
import net.nashihara.naroureader.utils.DownloadUtils
import net.nashihara.naroureader.views.RankingRecyclerView
import net.nashihara.naroureader.widgets.FilterDialogFragment
import net.nashihara.naroureader.widgets.ListDialogFragment
import net.nashihara.naroureader.widgets.NovelDownloadDialogFragment
import net.nashihara.naroureader.widgets.OkCancelDialogFragment

import java.util.ArrayList
import java.util.Arrays

import narou4j.entities.Novel
import narou4j.enums.NovelGenre

class RankingRecyclerViewFragment : Fragment(), RankingRecyclerView {

    private lateinit var binding: FragmentRankingRecyclerBinding

    private var replaceListener: FragmentTransactionListener? = null

    private var allItems = ArrayList<NovelItem>()

    private lateinit var controller: RankingRecyclerPresenter

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        replaceListener = context as FragmentTransactionListener?
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate<FragmentRankingRecyclerBinding>(inflater!!, R.layout.fragment_ranking_recycler, container, false)

        controller = RankingRecyclerPresenter(this)

        binding.fab.setOnClickListener { setupFab() }

        binding.recycler.layoutManager = LinearLayoutManager(context)

        val adapter = setupRecyclerView()
        binding.recycler.adapter = adapter

        arguments?.let {
            val typeStr = it.getString(PARAM_TYPE, "")
            controller.fetchRanking(typeStr)
        }
        return binding.root
    }

    private fun setupFab() {
        val filters = ArrayList(Arrays.asList(*resources.getStringArray(R.array.genres)))
        filters.add(0, "完結済み")

        val filterIds = NovelGenre.values()


        val checked = BooleanArray(filters.size)
        checked[0] = false
        for (i in 1..checked.size - 1) {
            checked[i] = true
        }
        val checkBoxDialog = FilterDialogFragment.newInstance("小説絞込み", filters.toTypedArray(), checked, true,
                object : FilterDialogFragment.OnDialogButtonClickListener {
                    override fun onPositiveButton(which: Int, itemChecked: BooleanArray, min: String, max: String) {
                        controller.filterNovelRanking(allItems, filterIds, itemChecked, min, max)
                    }

                    override fun onNeutralButton(which: Int) {
                        val adapter = binding.recycler.adapter as NovelDetailRecyclerViewAdapter
                        adapter.list.clear()
                        adapter.list.addAll(allItems)
                    }
                })
        checkBoxDialog.show(fragmentManager, "multiple")
    }

    private fun setupRecyclerView(): NovelDetailRecyclerViewAdapter {
        val adapter = NovelDetailRecyclerViewAdapter(context, false)
        adapter.setOnItemClickListener(object : NovelDetailRecyclerViewAdapter.OnItemClickListener {
            override fun onItemClick(view: View, position: Int, itemBinding: ItemRankingRecyclerBinding) {
                if (view.id != R.id.btn_expand) {
                    val item = (binding.recycler.adapter as NovelDetailRecyclerViewAdapter).list.get(position)
                    replaceListener?.replaceFragment(NovelTableRecyclerViewFragment
                            .newInstance(item.novelDetail.ncode), item.novelDetail.title, item)
                    return
                }

                if (itemBinding.allStory.visibility == View.GONE) {
                    itemBinding.allStory.visibility = View.VISIBLE
                    itemBinding.keyword.visibility = View.VISIBLE
                    itemBinding.btnExpand.setImageResource(R.drawable.ic_expand_less_black_24dp)
                } else {
                    itemBinding.allStory.visibility = View.GONE
                    itemBinding.keyword.visibility = View.GONE
                    itemBinding.btnExpand.setImageResource(R.drawable.ic_expand_more_black_24dp)
                }
            }

            override fun onItemLongClick(view: View, position: Int, itemBinding: ItemRankingRecyclerBinding) {

                val adapter = binding.recycler.adapter as NovelDetailRecyclerViewAdapter

                val item = adapter.list.get(position)
                val strings = arrayOf("小説を読む", "ダウンロード", "ブラウザで小説ページを開く", "ブラウザで作者ページを開く")
                val listDialog = ListDialogFragment.newInstance(item.novelDetail.title, strings,
                        DialogInterface.OnClickListener { _, which ->  longClickListDialogListener(which, position, item)})
                listDialog.show(fragmentManager, "list_dialog")
            }
        })

        return adapter
    }

    private fun longClickListDialogListener(which: Int, position: Int, item: NovelItem) {
        when (which) {
            0 -> {
                val item1 = (binding.recycler.adapter as NovelDetailRecyclerViewAdapter).list.get(position)
                replaceListener?.replaceFragment(NovelTableRecyclerViewFragment
                        .newInstance(item1.novelDetail.ncode), item1.novelDetail.title, item1)
            }
            1 -> {
                val downloadUtils = object : DownloadUtils() {
                    override fun onDownloadSuccess(dialog: NovelDownloadDialogFragment, novel: Novel) {
                        dialog.dismiss()

                        val okCancelDialog = OkCancelDialogFragment.newInstance("ダウンロード完了", "ダウンロードした小説を開きますか？") { dialog1, which1 ->
                            dialog1.dismiss()
                            if (OkCancelDialogFragment.OK == which1) {
                                val novelItem = NovelItem()
                                novelItem.novelDetail = novel
                                replaceListener!!.replaceFragment(
                                        NovelTableRecyclerViewFragment.newInstance(novel.ncode),
                                        novelItem.novelDetail.title, novelItem)
                            }
                        }
                        okCancelDialog.show(fragmentManager, "okcansel")
                    }

                    override fun onDownloadError(dialog: NovelDownloadDialogFragment) {
                        Log.d(TAG, "onDownloadError: ")
                        dialog.dismiss()
                    }
                }
                downloadUtils.novelDownlaod(item.novelDetail, fragmentManager, context)
            }
            2 -> {
                val url = "http://ncode.syosetu.com/" + item.novelDetail.ncode + "/"
                val uri = Uri.parse(url)
                val intent = Intent(Intent.ACTION_VIEW, uri)
                startActivity(intent)
            }
            3 -> {
                val url = "http://mypage.syosetu.com/" + item.novelDetail.userId + "/"
                val uri = Uri.parse(url)
                val intent = Intent(Intent.ACTION_VIEW, uri)
                startActivity(intent)
            }
        }
    }


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
    }

    private fun reload() {
        fragmentManager.beginTransaction().detach(this).attach(this).commit()
    }

    private fun onLoadError() {
        binding.progressBar.visibility = View.GONE
        binding.btnReload.visibility = View.VISIBLE
        binding.btnReload.setOnClickListener {
            binding.progressBar.visibility = View.VISIBLE
            binding.btnReload.visibility = View.GONE
            reload()
        }
    }

    override fun showRanking(novelItems: List<NovelItem>) {
        if (allItems.isEmpty()) {
            allItems = ArrayList(novelItems)
        }

        val adapter = binding.recycler.adapter as NovelDetailRecyclerViewAdapter
        adapter.clearData()
        adapter.addDataOf(novelItems)
        binding.progressBar.visibility = View.GONE
        binding.recycler.visibility = View.VISIBLE
    }

    override fun showError() {
        onLoadError()
    }

    companion object {

        private val TAG = RankingRecyclerViewFragment::class.java.simpleName

        private val PARAM_TYPE = "rankingType"

        fun newInstance(type: String): RankingRecyclerViewFragment {
            val fragment = RankingRecyclerViewFragment()
            val args = Bundle()
            args.putString(PARAM_TYPE, type)
            fragment.arguments = args
            return fragment
        }
    }
}
